Bump build requirement to JDK 17

ppkarwasz opened the pull request:
This change:

- Bumps the build requirement for the project to JDK 17.
- Adds a `java8-tests` profile that runs tests using JRE 8. The profile activates if a Maven Toolchains configuration file is present, i.e., the user knows about Maven Toolchains. Toolchains are automatically configured by `actions/setup-java`.
- Bumps the version of Error Prone to a version compatible with JDK 17 (i.e., latest).
- Bumps the version of Maven Surefire to a version that supports the `jdkToolchain` property.
- Locks the version of the remaining plugins to the currently used versions.

jeremylong commented:
Thanks! However, for the the CI to work we also need to update https://github.com/package-url/packageurl-java/blob/dd8309bc0b70c60df471154813f41a6582887327/.github/workflows/maven.yml#L10-L11

ppkarwasz commented:
> Thanks! However, for the the CI to work we also need to update

Fixed in https://github.com/package-url/packageurl-java/pull/178/commits/5dd8b8fcdc11ec5e1a2fe179a94aef9eec20f061 . Since tests run using JRE 8, two JDKs must be installed (or JRE 8 and JDK 17 if your prefer).

> Any chance you could add a `.java-version` as well?

Fixed in https://github.com/package-url/packageurl-java/pull/178/commits/91d8d047e954bb7111d77007016ad4a5ef48e36c

dwalluck commented:
Just a note that the spotbugs plugin would have to be bumped to 4.9.0 to support JDK 23.

dwalluck commented on .java-version:
```diff
@@ -0,0 +1 @@
+17
```
This file is missing the final newline.

ppkarwasz replied:
Thanks, fixed in https://github.com/package-url/packageurl-java/pull/178/commits/da461439c37e610491c418117155a832af6bfc74.

ppkarwasz commented:
> Just a note that the spotbugs plugin would have to be bumped to 4.9.0 to support JDK 23.

I would bump all Maven plugins. Previously many plugin versions were not specified, so their versions were determined by the Maven Lifecycle Bindings and were different for each version of Maven.

In this PR I fixed the version of all plugins to those of the [Maven 3.8.7 Lifecycle Bindings](https://maven.apache.org/ref/3.8.7/maven-core/default-bindings.html) (Debian 12 ships with Maven 3.8.7), but these are **really** old versions.

dwalluck commented:
I created #180 to try to prevent this, but I am not sure if your editor supports it https://editorconfig.org/#pre-installed.

A missing newline in a text file actually violates the POSIX standard as far as I know.

ppkarwasz commented:
> I created #180 to try to prevent this, but I am not sure if your editor supports it https://editorconfig.org/#pre-installed.

Sure, I'll enable the EditorConfig plugin. For Java code, however, I would enforce a more precise formatting (see #183 ).

> A missing newline in a text file actually violates the POSIX standard as far as I know.

Yes it does.

jeremylong commented on pom.xml:
```diff
@@ -250,6 +337,41 @@
+          ~ It activates if a toolchains configuration file is present.
+          ~ See: https://maven.apache.org/guides/mini/guide-using-toolchains.html
+          -->
+        <profile>
+            <id>java8-tests</id>
+            <activation>
+                <file>
+                    <exists>${user.home}/.m2/toolchains.xml</exists>
+                </file>
+            </activation>
```
I do not understand this profile. If the build requires Java 17, how could this be activated and work? What am I missing?

ppkarwasz replied:
The purpose of this profile is to run tests using JRE 8.
Since this can be only done through [Maven Toolchains](https://maven.apache.org/guides/mini/guide-using-toolchains.html), which might be hard to configure for single-time contributors, the profile only activates if a toolchains configuration is present.

Note that the CI always has a toolchains configuration.

dwalluck replied:
I hadn't really used it either, but since GitHub automatically sets up the file, it's pretty painless.

I think this pull request makes it so that it's not required (presumably, the CI will test Java 8 if the developer does not), but if you're interested, it has a mojo to automatically try to generate the config file `mvn toolchains:generate-jdk-toolchains-xml`.

ppkarwasz replied:
Cool! :100:
It discovered all 8 JDKs installed on my system.
